sql/core/src/main/scala/org/apache/spark/sql/Column.scala

```diff
@@ -901,39 +901,125 @@ class Column(val expr: Expression) extends Logging {
    *   // result: org.apache.spark.sql.AnalysisException: Ambiguous reference to fields
    * }}}
    *
+   * This method supports adding/replacing nested fields directly e.g.
+   *
+   * {{{
+   *   val df = sql("SELECT named_struct('a', named_struct('a', 1, 'b', 2)) struct_col")
+   *   df.select($"struct_col".withField("a.c", lit(3)).withField("a.d", lit(4)))
+   *   // result: {"a":{"a":1,"b":2,"c":3,"d":4}}
+   * }}}
+   *
+   * However, if you are going to add/replace multiple nested fields, it is more optimal to extract
+   * out the nested struct before adding/replacing multiple fields e.g.
+   *
+   * {{{
+   *   val df = sql("SELECT named_struct('a', named_struct('a', 1, 'b', 2)) struct_col")
+   *   df.select($"struct_col".withField("a", $"struct_col.a".withField("c", lit(3)).withField("d", lit(4))))
+   *   // result: {"a":{"a":1,"b":2,"c":3,"d":4}}
+   * }}}
+   *
    * @group expr_ops
    * @since 3.1.0
    */
   // scalastyle:on line.size.limit
   def withField(fieldName: String, col: Column): Column = withExpr {
     require(fieldName != null, "fieldName cannot be null")
     require(col != null, "col cannot be null")
+    updateFieldsHelper(expr, nameParts(fieldName), name => WithField(name, col.expr))
+  }
 
-    val nameParts = if (fieldName.isEmpty) {
+  // scalastyle:off line.size.limit
+  /**
+   * An expression that drops fields in `StructType` by name.
```

Comment:
Let's explicitly mention that, if the name doesn't match any field, it's noop.

Reply:
I've made this change but now that I think about it, I don't think its actually classifies as a "noop". We still reconstruct the struct unfortunately e.g.
```
val structType = StructType(Seq(
    StructField("a", IntegerType, nullable = false),
    StructField("b", IntegerType, nullable = true),
    StructField("c", IntegerType, nullable = false)))

val structLevel1: DataFrame = spark.createDataFrame(
    sparkContext.parallelize(Row(Row(1, null, 3)) :: Nil),
    StructType(Seq(StructField("a", structType, nullable = false))))

structLevel1.withColumn("a", 'a.dropFields("d")).explain()

== Physical Plan ==
*(1) Project [named_struct(a, a#1.a, b, a#1.b, c, a#1.c) AS a#3]
+- *(1) Scan ExistingRDD[a#1]
```
Should I revert this?

Reply:
It's semantically noop. We can optimize away the struct reconstructing later.